### ISSUE TITLE
Specific click program hangs with python popen, others run smoothly

### ISSUE THREAD

**Issue by kthkaya** [kohler/click#313]
Hello,

I have a python program where I launch click codes with subprocess library's popen method. I have 5 different network functions implemented in click (router, firewall, napt etc.) and all run successfully with popen.

You may think this is not a click related question but I doubt I can find an answer for this in python related forums, so I appreciate any guidance.

I have the following sink.click code.

AddressInfo(SNK 192.168.1.12 42:cb:c6:76:a0:80);
DUMP            :: ToDump("sink.pcap");
ETH_CLS         :: Classifier(12/0806 20/0001,12/0800, -);
PKT_CLS         :: IPClassifier(icmp,udp,tcp,-);
IN                      :: FromDevice(eth0, METHOD LINUX);
OUT             :: Queue -> ToDevice(eth0);
ARPR            :: ARPResponder(SNK) -> OUT;
Socket(TCP,0.0.0.0,80) -> DUMP;

IN ->
        ETH_CLS
                ->ARPR;
        ETH_CLS[1] -> Strip(14) -> CheckIPHeader ->
                PKT_CLS
                        -> StoreICMPTimeSeqRecord(OFFSET 0, DELTA true)
                        -> Unstrip(14)
                        -> DUMP;
                PKT_CLS[1]
                        -> StoreUDPTimeSeqRecord(OFFSET 0, DELTA true)
                        -> Unstrip(14)
                        -> DUMP;
                PKT_CLS[2]
                        -> StoreTCPTimeSeqRecord(OFFSET 0, DELTA true)
                        -> Unstrip(14)
                        -> DUMP;
                PKT_CLS[3]
                        -> Discard;
        ETH_CLS[2]
                -> Discard;

Now when I run _click sink.click_ from the command line, it just works. It creates the sink.pcap file immediately. But when I try to run it from a python code like the following
`
subprocess.Popen(['click','sink.click'])
`
it launches, process is created and it is not blocking the python program that launches it. However it doesn't do its function, doesn't even create the blank pcap file.
Note that I am using the same Popen method for my other click codes and all work! Only difference I can think of between this code and the others, is that this is the only click code that creates a socket. 

Any thoughts would be highly appreciated

**Comment by tbarbette** [comment]
Hi,
This code works for me with Pyton 3 (remove the x.decode thing for python 2) : 
It tries to find a "RESULT [0-9.]" line after execution and return that, stderr and stdout.
As I do not install Click globally, I also update the PATH environment variable to find Click.

`p=Popen(self.script.content,stdin=PIPE,stdout=PIPE,stderr=PIPE,shell=True,env={"PATH":self.appdir + build.repo.reponame+"/build/bin:" + os.environ["PATH"]})
            try:
                output, err = [x.decode() for x in p.communicate(self.stdin.content, timeout=self.config["timeout"])]
                nr = re.search("RESULT ([0-9.]+)",output.strip())

                if (nr):
                    n = float(nr.group(1))
                    results.append(n)
                else:
                    print("Could not find result !")
                    print("stdout:")
                    print(output)
                    print("stderr:")
                    print(err)
                    return False, output, err
            except TimeoutExpired:
                print("Test expired")
                return False,"Timeout expired.",""`

**Comment by kthkaya** [comment]
Thanks for your response Tom. I used your code in a fresh python file and it worked. Then I compared with my existing code and it turned out the pcap file was also being created in my code, but not in the folder I expected as I haven't given full path for the pcap in click code. When you run the command with click binary, the pcap is created in the same folder that the click file exists. But when you run with python (which does nothing but issue the same command), it is created where the python file is.

Been struggling with this stupid mistake for quite some time now and your input helped me approach from scratch with fresh eyes. Really appreciated!
